File: Scala/src/main/scala/org/revian/MyQueue.scala
package org.revian

class MyQueue(capacity: Int) {
  private var queue = List[Int]()
  private val maxCapacity = capacity

  def enqueue(value: Int): Unit = {
    if (queue.size >= maxCapacity) {
      throw new IllegalStateException("Queue is full")
    }
    queue = queue :+ value
  }

  def dequeue(): Int = {
    if (queue.isEmpty) throw new IllegalStateException("Queue is empty")
    val first = queue.head
    queue = queue.tail
    first
  }

  def size(): Int = queue.length
}

File: Java/src/main/java/org/revian/MyQueue.java
package org.revian;

public class MyQueue {
    private int[] queue;
    private int front;
    private int rear;
    private int size;
    private static final int DEFAULT_CAPACITY = 10;

    public MyQueue() {
        queue = new int[DEFAULT_CAPACITY];
        front = 0;
        rear = 0;
        size = 0;
    }

    public void enqueue(int value) {
        if (size == queue.length) {
            throw new IllegalStateException("Queue is full");
        }
        queue[rear] = value;
        rear = (rear + 1) % queue.length;
        size++;
    }

    public int dequeue() {
        if (size == 0) {
            throw new IllegalStateException("Queue is empty");
        }
        int value = queue[front];
        front = (front + 1) % queue.length;
        size--;
        return value;
    }

    public int size() {
        return size;
    }
}
